mautic/mautic
# API additions

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | y
| Related user documentation PR URL | 
| Related developer documentation PR URL | todo
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | y

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

This PR has the following as apart of it due to needing shared code changes:

[ ] #3193

[//]: # ( Required: )
#### Description:

This PR started out with just adding campaign API endpoints to edit/schedule campaign events for contacts. But then it grew into more due to shared code between new features

1. Adds ability to edit or schedule an individual campaign event for a specific contact via the API. Also adds the ability to get campaign event logs for a specific contact or campaign. 
2. Adds user permission filtering to the stats endpoints.
3. Added batch support for API write endpoints
4. Random but adds option to edit or cancel scheduled campaign events through the contact timeline.

#### Steps to test this PR:
1. Run the tests in the associated API PR - https://github.com/mautic/api-library/pull/79
2. For #4 - run a campaign that schedules events. View the details of a contact and filter the events by "Campaign event scheduled." Click the icon to show the details and notice there's two new buttons. Click the clock button to edit the time and/or the cancel button to cancel the event. 

#### List deprecations along with the new alternative:
1. The following API urls are deprecated to create a consistent structure across endpoints:
* `/campaigns/{id}/contact/add/{leadId}` -> `/campaigns/{id}/contact/{leadId}/add`
* `/campaigns/{id}/contact/remove/{leadId}` -> `/campaigns/{id}/contact/{leadId}/remove`
* `/emails/{id}/send/contact/{leadId}` -> `/emails/{id}/contact/{leadId}/send`
* `/segments/{id}/contact/add/{leadId}` -> `/segments/{id}/contact/{leadId}/add`
* `/segments/{id}/contact/remove/{leadId}` -> `/segments/{id}/contact/{leadId}/remove`
* `/companies/{companyId}/contact/add/{contactId}` ->  `/companies/{companyId}/contact/{contactId}/add`
* `/companies/{companyId}/contact/remove/{contactId}` -> `/companies/{companyId}/contact/{contactId}/remove`
* `/contacts/{id}/dnc/add/{channel}`  -> `/contacts/{id}/dnc/{channel}/add`
* `/contacts/{id}/dnc/remove/{channel}` -> `/contacts/{id}/dnc/{channel}/remove`
* `/stages/{id}/contact/add/{contactId}` -> `/stages/{id}/contact/{contactId}/add`
* `/stages/{id}/contact/remove/{contactId}` -> `/stages/{id}/contact/{contactId}/remove`
